rust/kernel/thread.rs

```diff
@@ -0,0 +1,404 @@
+    /// # Memory Cost
+    ///
+    /// The implementation of this function currently needs to allocate extra
+    /// dynamic memory (using [`Box`]), and do double boxing because
+    /// `Box<dyn FnOnce()>`. These "drawbacks" similarly exist in the
+    /// implementation of [`std::thread::spawn`]. Only use this when these
+    /// are acceptable, otherwise try [`Thread::try_new`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use alloc::boxed::Box;
+    /// use kernel::thread::Thread;
+    ///
+    /// let mut a = 1;
+    /// let boxed_slice: Box<[i32]> = Box::try_new([1, 3, 4, 5])?; // Closures can capture DSTs.
+    ///
+    /// let t = Thread::try_new_closure(cstr!("rust-thread"), move || {
+    ///     let b = Box::try_new(42)?;
+    ///
+    ///     for _ in 0..10 {
+    ///         a = a + 1;
+    ///         println!("Hello Rust Thread {}", a + b.as_ref());
+    ///     }
+    ///
+    ///     println!("A Box<[i32]>: {:?}", boxed_slice);
+    ///     Ok(())
+    /// })?;
+    ///
+    /// t.wake_up();
+    /// ```
+    ///
+    /// # Context
+    ///
+    /// This function might sleep in `kthread_create_on_node` due to the memory
+    /// allocation and waiting for the completion, therefore do not call this
+    /// in atomic contexts (i.e. preemption-off contexts).
+    ///
+    /// [`std::thread::spawn`]: https://doc.rust-lang.org/std/thread/fn.spawn.html
+    pub fn try_new_closure<F>(name: CStr, f: F) -> KernelResult<Self>
+    where
+        F: FnOnce() -> KernelResult,
+        F: Send + 'static,
+    {
+        // Allocate closure here, because this function maybe returns before
+        // the bridged function (the function that uses the closure) get executed.
+        let boxed_fn: Box<dyn FnOnce() -> KernelResult + Send> = Box::try_new(f)?;
+
+        // Double boxing here because `dyn FnOnce` is a fat pointer, and we cannot
+        // `into_pointer` it to `*mut c_void`, see comments of `ClosureCallFunctor`
+        // for more.
+        let double_box = Box::try_new(boxed_fn)?;
+
+        Self::try_new::<_, ClosureCallFunctor>(name, double_box)
+    }
+
+    /// Wakes up the thread.
+    ///
+    /// Note that a newly created thread (e.g. via [`Thread::try_new`]) will not
+    /// run until a [`Thread::wake_up`] is called.
+    ///
+    /// # Context
+    ///
+    /// This function might sleep, don't call it in atomic contexts.
+    pub fn wake_up(&self) {
+        // SAFETY: `task` is a valid pointer to a kernel thread structure, the refcount
+        // of which is increased in `try_new*`, so it won't point to a freed
+        // `task_struct`. And it's not stopped because `stop` will consume the
+        // [`Thread`].
+        unsafe {
+            bindings::wake_up_process(self.task);
+        }
+    }
+
+    /// Stops the thread.
+    ///
+    /// - If the thread hasn't been waken up after creation, the thread closure
+    ///   won't be called, and will return `-EINTR`. Note that a thread may not
```

Comment:
Does it mean that we may leak memory because no one would do `from_pointer`?

Reply:
I think you're right! So this is a bug, we can fix this by introducing another field in `struct Thread` to hold the a pointer to the thread arg, if `EINTR` is returned by `kthread_stop`, we should `from_pointer` in the `stop`.

Reply:
What if the thread itself returns EINTR? Then we could potentially double-free.

Reply:
We can disallow a thread doing that. 

There are two approaches that a thread sets its exit code: 1) via `do_exit` and 2) via thread function return value, for 1) we can provide Rust version of `do_exit` and change `-EINTR` into other value (actually, no `do_exit` caller uses `-EINTR` currently in kernel, except `kthread` function itself), for 2) since we provide the thread function (the `bridge` function), we can rewrite the return value of thread function if it's `-EINTR`. This may not be elegant, but a kernel exits itself with `-EINTR` also looks very strange to me.

Reply:
I've looked into this, here are some of my findings and thoughts:

1. If a thread is created by `try_new_extern`, we actually don't need to `from_pointer` in `stop`, because `try_new_extern` users are supposed to handle the arguments themselves.
2. Since neither `Thread` nor `stop` is generic, we actually cannot handle the threads created by `'try_new`, because we don't know the types we want `from_pointer` to. 
    1. Making `stop` to be a generic function seems also implying `unsafe`, because the caller need to guarantee the type is correct. 
    2. Making `Thread` generic seems challenging, because Rust doesn't allow unused generic parameters in structures. I will think more about this.
3. The handing of threads created by `try_new_closure` is rather easy, because the thread arguments have the same type.

Looks to me, the feasible way is creating three thread structures (differed by the constructors), and handle them separately. Thoughts

Reply:
> 1. If a thread is created by `try_new_extern`, we actually don't need to `from_pointer` in `stop`, because `try_new_extern` users are supposed to handle the arguments themselves.

Agreed.

> 2. Since neither `Thread` nor `stop` is generic, we actually cannot handle the threads created by `try_new`, because we don't know the types we want `from_pointer` to.
>    
>    1. Making `stop` to be a generic function seems also implying `unsafe`, because the caller need to guarantee the type is correct.
>    2. Making `Thread` generic seems challenging, because Rust doesn't allow unused generic parameters in structures. I will think more about this.

Can't you use `core::marker::PhantomData` for this?

Reply:
> >    2. Making `Thread` generic seems challenging, because Rust doesn't allow unused generic parameters in structures. I will think more about this.
> 
> Can't you use `core::marker::PhantomData` for this?

Clearly I wasn't aware that I could use it ;-) Thanks for the suggestion!